[Enhancement] Improvements for debounce test coverage + bug fixes for sym_defer_g and sym_eager_pr

andrebrait opened the pull request:
## Description

* Add tests for none
* Add check for consistency between cooked_changed and actual result
* Add check for ensuring debounce never reads the timer more than once
* Fix sym_eager_pr incorrectly returning cooked_changed as false
* Fix sym_defer_g reading the timer twice when debouncing
* Save some useless memory copies on none when there are no changes
* Restructure sym_defer_g so it is smaller on AVR

<!--- Describe your changes in detail here. -->

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [X] Core
- [X] Bugfix
- [ ] New feature
- [X] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

andrebrait commented:
@lesshonor thanks. Brainfart. I rebased and changed the target branch.

Nebuleon commented on quantum/debounce/sym_defer_pk.c:
```diff
@@ -56,13 +56,9 @@ static void start_debounce_counters(matrix_row_t raw[], matrix_row_t cooked[], u
-            debounce_counters[i++] = DEBOUNCE_ELAPSED;
-        }
-    }
+    const size_t debounce_counters_size = num_rows * MATRIX_COLS * sizeof(debounce_counter_t);
+    debounce_counters                   = (debounce_counter_t *)malloc(debounce_counters_size);
+    memset(debounce_counters, DEBOUNCE_ELAPSED, debounce_counters_size);
```
This `memset` can easily break if this code gets revisited later on. If the type of `debounce_counter_t` gets upgraded at the same time as the value of `DEBOUNCE_ELAPSED` becomes non-zero, both of which are abstracted and which a developer will reasonably assume can be changed without much fuss, the counters will then get initialized to the wrong value.
```suggestion
    debounce_counters = (debounce_counter_t *)malloc(num_rows * MATRIX_COLS * sizeof(debounce_counter_t));
    for (size_t i = 0; i <= num_rows * MATRIX_COLS; i++) {
        debounce_counters[i] = DEBOUNCE_ELAPSED;
    }
```

andrebrait replied:
I'm not sure I understand what you mean. Unless debounce_counters changes to being a pointer of pointers?

Nebuleon replied:
I meant more like changing it to `uint16_t`. Since `memset` works on `char`, if the type `debounce_counter_t` gets changed to `uint16_t`, with a `DEBOUNCE_ELAPSED` value of 1, `memset` will make the elements 257 (0x0101) rather than 1.

andrebrait replied:
Got it. I will change it to a loop then.

Nebuleon commented on quantum/debounce/sym_defer_pr.c:
```diff
@@ -21,51 +21,75 @@ DEBOUNCE milliseconds have elapsed since the last change.
 void debounce_init(uint8_t num_rows) {
-    countdowns = (uint8_t*)calloc(num_rows, sizeof(uint8_t));
-    last_raw   = (matrix_row_t*)calloc(num_rows, sizeof(matrix_row_t));
+    const size_t debounce_counters_size = num_rows * sizeof(debounce_counter_t);
+    debounce_counters                   = (debounce_counter_t*)malloc(debounce_counters_size);
+    memset(debounce_counters, DEBOUNCE_ELAPSED, debounce_counters_size);
```
This `memset` can easily break if this code gets revisited later on. If the type of `debounce_counter_t` gets upgraded at the same time as the value of `DEBOUNCE_ELAPSED` becomes non-zero, both of which are abstracted in this commit and which a developer will reasonably assume can be changed without much fuss, the counters will then get initialized to the wrong value.
```suggestion
    debounce_counters = (debounce_counter_t*)malloc(num_rows * sizeof(debounce_counter_t));
    for (uint8_t r = 0; r <= num_rows; r++) {
        debounce_counters[r] = DEBOUNCE_ELAPSED;
    }
```

Nebuleon commented on quantum/debounce/sym_defer_pr.c:
```diff
@@ -21,51 +21,75 @@ DEBOUNCE milliseconds have elapsed since the last change.
-    countdowns = (uint8_t*)calloc(num_rows, sizeof(uint8_t));
-    last_raw   = (matrix_row_t*)calloc(num_rows, sizeof(matrix_row_t));
+    const size_t debounce_counters_size = num_rows * sizeof(debounce_counter_t);
+    debounce_counters                   = (debounce_counter_t*)malloc(debounce_counters_size);
```
Inconsistent formatting of pointer types between changes introduced in this file (`debounce_counter_t*`) and in `quantum/debounce/sym_defer_pk.c` (`debounce_counter_t *`). [Please run `clang-format`](https://docs.qmk.fm/#/coding_conventions_c?id=auto-formatting-with-clang-format) to lint your files.

This does mean that, while some of my suggestions can be applied directly on GitHub, they will need more attention, as I have kept your formatting in each instance.

Nebuleon commented on quantum/debounce/sym_eager_pk.c:
```diff
@@ -57,13 +57,9 @@ static void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], ui
-            debounce_counters[i++] = DEBOUNCE_ELAPSED;
-        }
-    }
+    const size_t debounce_counters_size = num_rows * MATRIX_COLS * sizeof(debounce_counter_t);
+    debounce_counters                   = (debounce_counter_t *)malloc(debounce_counters_size);
+    memset(debounce_counters, DEBOUNCE_ELAPSED, debounce_counters_size);
```
This `memset` can easily break if this code gets revisited later on. If the type of `debounce_counter_t` gets upgraded at the same time as the value of `DEBOUNCE_ELAPSED` becomes non-zero, both of which are abstracted in this commit and which a developer will reasonably assume can be changed without much fuss, the counters will then get initialized to the wrong value.
```suggestion
    debounce_counters = (debounce_counter_t *)malloc(num_rows * MATRIX_COLS * sizeof(debounce_counter_t));
    for (size_t i = 0; i <= num_rows * MATRIX_COLS; i++) {
        debounce_counters[i] = DEBOUNCE_ELAPSED;
    }
```

Nebuleon commented on quantum/debounce/sym_eager_pr.c:
```diff
@@ -56,12 +56,10 @@ static void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], ui
+    debounce_counters                   = (debounce_counter_t *)malloc(debounce_counters_size);
+    memset(debounce_counters, DEBOUNCE_ELAPSED, debounce_counters_size);
 }
-
```
Please add back this blank line between functions.

Nebuleon commented on quantum/debounce/sym_eager_pr.c:
```diff
@@ -56,12 +56,10 @@ static void transfer_matrix_values(matrix_row_t raw[], matrix_row_t cooked[], ui
-    for (uint8_t r = 0; r < num_rows; r++) {
-        debounce_counters[r] = DEBOUNCE_ELAPSED;
-    }
+    const size_t debounce_counters_size = num_rows * sizeof(debounce_counter_t);
+    debounce_counters                   = (debounce_counter_t *)malloc(debounce_counters_size);
+    memset(debounce_counters, DEBOUNCE_ELAPSED, debounce_counters_size);
```
This `memset` can easily break if this code gets revisited later on. If the type of `debounce_counter_t` gets upgraded at the same time as the value of `DEBOUNCE_ELAPSED` becomes non-zero, both of which are abstracted in this commit and which a developer will reasonably assume can be changed without much fuss, the counters will then get initialized to the wrong value.

The previous code was good in that respect.
```suggestion
    debounce_counters = (debounce_counter_t *)malloc(num_rows * sizeof(debounce_counter_t));
    for (uint8_t r = 0; r < num_rows; r++) {
        debounce_counters[r] = DEBOUNCE_ELAPSED;
    }
```

filterpaper commented on quantum/debounce/sym_defer_g.c:
```diff
@@ -24,23 +24,31 @@ When no state changes have occured for DEBOUNCE milliseconds, we push the state.
     if (changed) {
         debouncing      = true;
         debouncing_time = timer_read_fast();
+        return false;
```
This `return false` statement will add quite a number of bytes for AVR and it is the _default_ debounce algorithm. I'm curious how it will prevernt sym_defer_g from possibly skipping debounce.

Nebuleon replied:
The logic seems to be that if the debounce algorithm has just set the `debouncing_time` variable, it can't be time to actually debounce yet, especially since the only information available to debounce is the matrix contents in `raw` and `cooked` from the same invocation of `debounce`, and we need to wait for the next one. In that sense, this is a CPU time optimization.

Applying just the changes in this function to sofle:default, which uses `sym_defer_g` on AVR, causes an increase of 2 bytes.

andrebrait replied:
Exactly. I can move things around to try and make it smaller again. Also, combined with the changes to the TIME_DIFF macros, what is the total diff in size? I can't check right now. I got rid of a lot of branching there.

filterpaper replied:
Thanks for the clarification, so it is not feature change but cycle optimisation. Its impact approximately 8 bytes on other boards, such as `ferris/sweep`. I presume the _optimisation_ is just one single tick / cycle?

andrebrait replied:
Basically, I don't want to call the timer a second time and there's nothing that guarantees that we won't fall on some weird corner case scenario where the first timer call to set the `debouncing_time` will return exactly the same value as the second one, and that we won't accidentally skip debouncing when e.g. DEBOUNCE is set to 1.

Nebuleon replied:
Applying the entire PR onto current `develop` is also a 2 byte increase on sofle:default (with `sym_defer_g`). So there doesn't seem to be something else optimizing the size in the rest of the code to balance out the increase. The branching in `TIMER_DIFF_*` already got optimized away because the elements of the computation that determines the requested bit count are compile-time constants.

Nebuleon replied:
> [@filterpaper] Thanks for the clarification, so it is not feature change but cycle optimisation. Its impact approximately 8 bytes on other boards, such as `ferris/sweep`. I presume the _optimisation_ is just one single tick / cycle?

The code that gets skipped is as follows:
```c
    if (debouncing && timer_elapsed_fast(debouncing_time) >= DEBOUNCE) {
        if (memcmp(cooked, raw, sizeof(matrix_row_t) * num_rows) != 0) {
            memcpy(cooked, raw, sizeof(matrix_row_t) * num_rows);
            cooked_changed = true;
        }
        debouncing = false;
    }
```
Depending on the matrix's row count, avoiding `memcmp` and `memcpy` can be big.

andrebrait replied:
I'll see if I can manage to:

1. revert some of the changes/get rid of memset
2. make this smaller by replacing the separate return with an "else if" and keeping a single return statement

Either way, unless we can *always* guarantee in *all* architectures that two subsequent calls to timers will get the same value within this function (and I guess we can't, because ARM, for example (I'm almost certain?), can have the timer pass maybe 1ms in this time (between the latest nanos from one ms to the first nanos of the next, considering it's all integers)... we need to change this logic in order to ensure it absolutely never skips debounce.

filterpaper replied:
> The code that gets skipped is as follows:

"`debouncing_time = timer_read_fast();` should cause `timer_elapsed_fast(debouncing_time) >= DEBOUNCE` to evaluate `false` within the same _tick_?

andrebrait replied:
> > The code that gets skipped is as follows:
> 
> "`debouncing_time = timer_read_fast();`" should cause "`timer_elapsed_fast(debouncing_time) >= DEBOUNCE`" to evaluate `false` within the same _tick_?

Guaranteed to be the case in all platforms, including future ones?

I'd rather get the time once and use it twice with TIME_DIFF than to call thing sort of thing twice.

filterpaper replied:
> Guaranteed to be the case in all platforms, including future ones?

That was my original curiosity and technical interest — how or what platform will it be not guaranteed?

Nebuleon replied:
An interrupt could also happen between `debouncing_time = ...;` and `timer_elapsed_fast(debouncing_time)` in the same invocation of `debounce`, which would trigger the debounce to happen right then and there (with the same data, not a fresh matrix report) if the interrupt took `DEBOUNCE` ms.

There is an argument to be made for `if (changed) debouncing = true; else ...` being **more correct** than the current code. We **need** to get out as soon as possible so that a new matrix report can be given to a second invocatien of `debounce`.

In my opinion, that is worth making - and *thoroughly testing* - a second PR for, in all debounce implementations.

andrebrait replied:
Even if we currently have no platforms in which that could happen, the algorithm itself shouldn't be susceptible to such scenario. If we ever port QMK to a platform in which this is possible, it completely breaks. I think ARM could even be such a case right now, since the clock is derived from CPU cycles? Please correct me if I'm wrong.

Regarding the other algorithms, @Nebuleon, at least the `_pk` ones aren't affected, _I think_. First and foremost because they only invoke the timer once and then calculate the difference using `TIME_DIFF_FAST`, so no multiple invocations would take place, and since the timer is per key, it can't trigger a premature report. Not from what I checked, that is. I didn't check much about `_pr` though, but I did make some changes to one of them.

filterpaper replied:
> Applying just the changes in this function to sofle:default, which uses `sym_defer_g` on AVR, causes an increase of 2 bytes.

It will be worth pointing out that AVR boards close to size limit have LTO enabled. The byte difference with link time optimisation seems large large and non-trivial:

```c
qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
 * The firmware size is fine - 16742/28672 (58%, 11930 bytes free)
```
With the `return false` statement:
```c
qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
 * The firmware size is fine - 16790/28672 (58%, 11882 bytes free)
```

Nebuleon replied:
Eep, that's a lot of bytes indeed...

filterpaper replied:
The size increase does not impact all equally. There might be something else influencing the change.

andrebrait replied:
Running 

```bash
qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
```

using arm-none-eabi-gcc.exe (GCC) 12.2.0, I get:

* Original code:
```
* The firmware size is fine - 16692/28672 (58%, 11980 bytes free)
```

* With `const size_t matrix_size`:
```
 * The firmware size is fine - 16692/28672 (58%, 11980 bytes free)
 ```

* With an 'else if' + `const size_t matrix_size`:
```
* The firmware size is fine - 16700/28672 (58%, 11972 bytes free)
```

* With a `return false` + `const size_t matrix_size`:
```
* The firmware size is fine - 16700/28672 (58%, 11972 bytes free)
```

Nebuleon replied:
Using what version of `avr-gcc`? `cradio` uses an AVR atmega32u4 CPU.

andrebrait replied:
```
avr-gcc.exe (GCC) 12.2.0
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

I think it's whatever is right now on MSYS2 (I use QMK MSYS and I ran `pacman -Syuu` before these tests).

filterpaper replied:
`avr-gcc` 8.5 is recommended version installed by the script and used by the CI image because it produces the smallest binary.

That said, this looks like a compiler optimisation problem and the CI build passed for everything; so it should be immaterial for this PR.

andrebrait replied:
> `avr-gcc` 8.5 is the recommended version installed by the script and used by the CI image because it produces the smallest binary.
> 
> That said, this looks like a compiler optimisation problem and the CI build passed for everything; so it should be immaterial for this PR.

But my binaries with avr-gcc 12.2.0 are smaller than the ones generated by CI, aren't they? What version did you test that with?

filterpaper replied:
Your default binaries, without changes from this PR, is larger than the ones built by v8.5. That's the version I used for tests.

andrebrait replied:
Ok, I gotta check it again with 8.5, because with 12.2.0, I got 16692 bytes without the return, while you got 16742. And I got 16700 with the return, whereas you said you got 16790.

Either those numbers are wrong or 12.2.0 is generating smaller binaries.

filterpaper replied:
My apologies, I was reading them incorrectly (before coffee) as firmware output size, instead of free bytes. I think the smaller size optimisation is not true for every AVR board in the repository.

andrebrait replied:
After I rebased onto develop, the sizes became a tad larger. No idea why, though.

andrebrait replied:
Now with a clean install of QMK MSYS and `avr-gcc 8.5.0`:

With this PR:

```
[andre@Andre-PC qmk_firmware]$ git checkout 32b041900af5f0405bd32e382a89504c7d201a0b && make clean && make git-submodule
[andre@Andre-PC qmk_firmware]$ qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
 * The firmware size is fine - 16814/28672 (58%, 11858 bytes free)
 ```
 
 Without this PR:
 
 ```
[andre@Andre-PC qmk_firmware]$ git checkout 2e679450336b6679a7cfa4b024d33a10565b2e6 && make clean && make git-submodule
[andre@Andre-PC qmk_firmware]$ qmk compile -c -kb cradio -km default -e LTO_ENABLE=yes
* The firmware size is fine - 16766/28672 (58%, 11906 bytes free)
 ```
 
So yes, there's some size increase involved in the PR as a whole.
 
But, replacing `sym_defer_g.c` with the following contents, I got `16774 bytes`, which is a lot smaller. It seems that the call to the inlined function requires more space than the extra variable.

```c
#include "debounce.h"
#include "timer.h"
#include <string.h>
#ifndef DEBOUNCE
#    define DEBOUNCE 5
#endif

// Maximum debounce: 255ms
#if DEBOUNCE > UINT8_MAX
#    undef DEBOUNCE
#    define DEBOUNCE UINT8_MAX
#endif

#if DEBOUNCE > 0
static bool         debouncing = false;
static fast_timer_t debouncing_time;

#define NOW() (now > 0 ? now : timer_read_fast())

void debounce_init(uint8_t num_rows) {}

bool debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
    bool cooked_changed = false;
    fast_timer_t now = 0;

    if (changed) {
        debouncing      = true;
        debouncing_time = now = timer_read_fast();
    }

    if (debouncing && TIMER_DIFF_FAST(NOW(), debouncing_time) >= DEBOUNCE) {
        const size_t matrix_size = num_rows * sizeof(matrix_row_t);
        if (memcmp(cooked, raw, matrix_size) != 0) {
            memcpy(cooked, raw, matrix_size);
            cooked_changed = true;
        }
        debouncing = false;
    }

    return cooked_changed;
}

void debounce_free(void) {}
#else // no debouncing.
#    include "none.c"
#endif
```

filterpaper replied:
> But with [25bba8d](https://github.com/qmk/qmk_firmware/commit/25bba8d2807b28518ef251dcb57180d984fcd2e3) I got `16774 bytes`, which is a lot smaller.

I think it is because the `return false` statement is not within that `(changed)` block.

andrebrait replied:
I need to check exactly why. Using `changed` again, at all (like `&& !changed` or `else if`, etc.) has some weird side effect and the resulting image grows again, whereas storing `now` and selecting what to use based only on it doesn't result in increased size.

andrebrait replied:
I also need to code the test case to prevent this regressing in the future

andrebrait replied:
I marked this as resolved. If needed, let's start a new thread.

andrebrait commented:
Weird, I'm running `qmk format-c --core-only quantum/debounce/sym_defer_pr.c` but I'm getting something the CI doesn't consider food enough?

@drashna any tips here?

andrebrait commented:
Ok, everything is done now. Tests have been added, too.

There was one *big* omission from the tests in that they didn't check if the boolean returned from debounce actually matched when the output had not changed. When it returned `false`, the check for the equality with the output was skipped. Now it will ensure both scenarios are consistent. With that added, I managed to confirm the bug in `sym_eager_pr` (it would return `false` even when it had changed the cooked matrix, thus breaking the test) and fixed it in this PR as well.

I made a test for reproducing an async tick and, sure enough, `sym_global_g` broke. It's also fixed in the PR.

@filterpaper I managed to get the size down to 16768 bytes.

With some other changes, I can manage to get to 16752, by applying the patch below. Let me know if you think it's worth it. I honestly think this is even better, but I want your opinion on it.

```diff
diff --git a/platforms/avr/timer.c b/platforms/avr/timer.c
index 65b2db1343..ca89b3dc59 100644
--- a/platforms/avr/timer.c
+++ b/platforms/avr/timer.c
@@ -83,13 +83,7 @@ inline void timer_clear(void) {
  * FIXME: needs doc
  */
 inline uint16_t timer_read(void) {
-    uint32_t t;
-
-    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
-        t = timer_count;
-    }
-
-    return (uint16_t)t;
+    return (uint16_t)timer_read32();
 }

 /** \brief timer read32
@@ -111,27 +105,16 @@ inline uint32_t timer_read32(void) {
  * FIXME: needs doc
  */
 inline uint16_t timer_elapsed(uint16_t last) {
-    uint32_t t;
-
-    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
-        t = timer_count;
-    }
-
-    return TIMER_DIFF_16((uint16_t)t, last);
+    return TIMER_DIFF_16(timer_read(), last);
 }

 /** \brief timer elapsed32
  *
+
  * FIXME: needs doc
  */
 inline uint32_t timer_elapsed32(uint32_t last) {
-    uint32_t t;
-
-    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
-        t = timer_count;
-    }
-
-    return TIMER_DIFF_32(t, last);
+    return TIMER_DIFF_32(timer_read32(), last);
 }
 ```

andrebrait commented:
> 
> With some other changes, I can manage to get to 16752, by applying the patch below. Let me know if you think it's worth it. I honestly think this is even better, but I want your opinion on it.
> 

@filterpaper I went ahead and applied this patch. I can't see any downsides, with everything inlined.

andrebrait commented:
Ok, so, after digging a little bit further into the subject, I see a source of problems here.

It seems that, for AVR, we get the time from the usual means, using ISRs and getting the interrupt from the CPU and using a prescaler to get the interrupt to represent 1ms. This means, @Nebuleon and @filterpaper that I think it's also possible to get a jump of 1ms between subsequent invocations of a timer read inside the same call to debounce, even on AVR.

And due to integer division, we have a timing problem, because the current checks are basically (in algorithms) `timer_elapsed >= DEBOUNCE`, and that `>=` will possibly be true anywhere starting from `DEBOUNCE - 1 + (1 clock cycle above the prescaler)`, not at `DEBOUNCE` milliseconds or greater.

As an example (frequency of 16MHz and prescaler of 64, getting us 1 interrupt every 1ms, or 1 interrupt every 250 CPU cycles, and a `DEBOUNCE` of 5).

1. We check the timer at clock `x % 250 = 230` getting (for example) a value of 10. 
2. We run the debounce and store the time 10.
3. On another matrix update, we read the timer at exactly `x + 4 * 250 + 10` clock cycles, meaning we get 15 as return.
4. We consider `15 >= 10 + 5`, and send the matrix update.
5. The total debouncing time was thus 4.01ms, not 5.

We finished debouncing at less than 5 ms here. Currently, our approach only guarantees the debounce will happen at `>= DEBOUNCE -1` time.

So we have two choices here:

1. Move the checks back to `>`
2. Let go of the timer being in ms and count in either us or ns, and bit-shift the value gotten from the interrupts if necessary, and then the check becomes more precise since we'll have the actual time. 32-bit should be enough for the purposes of debouncing, so even in ns precision we're talking about 4 seconds before it rolls over.

@drashna not sure if this needs to involve other teams or something. Tagging you because you usually look at my PRs 😜

andrebrait commented:
Ok, so, after digging a little bit further into the subject, I see a source of problems here.

It seems that, for AVR, we get the time from the usual means, using ISRs and getting the interrupt from the CPU and using a prescaler to get the interrupt to represent 1ms. This means, @Nebuleon and @filterpaper that I think it's also possible to get a jump of 1ms between subsequent invocations of a timer read inside the same call to debounce, even on AVR.

And due to integer division, we have a timing problem, because the current checks are basically (in algorithms) `timer_elapsed >= DEBOUNCE`, and that `>=` will possibly be true anywhere starting from `DEBOUNCE - 1 + (1 clock cycle above the prescaler)`, not at `DEBOUNCE` milliseconds or greater.

As an example (frequency of 16MHz and prescaler of 64, getting us 1 interrupt every 1ms, or 1 interrupt every 250 CPU cycles, and a `DEBOUNCE` of 5).

1. We check the timer at clock `x % 250 = 230` getting (for example) a value of 10. 
2. We run the debounce and store the time 10.
3. On another matrix update, we read the timer at exactly `x + 4 * 250 + 10` clock cycles, meaning we get 15 as return.
4. We consider `15 >= 10 + 5`, and send the matrix update.
5. The total debouncing time was thus 4.01ms, not 5.

We finished debouncing at less than 5 ms here. Currently, our approach only guarantees the debounce will happen at `>= DEBOUNCE -1` time.

So we have two choices here:

1. Move the checks back to `>`
2. Let go of the timer being in ms and count in either us or ns, and bit-shift the value gotten from the interrupts if necessary, and then the check becomes more precise since we'll have the actual time. 32-bit should be enough for the purposes of debouncing, so even in ns precision we're talking about 4 seconds before it rolls over.

@drashna not sure if this needs to involve other teams or something. Tagging you because you usually look at my PRs 😜

sigprof commented:
The problem with making the debounce timers more precise is that the RAM requirements for the per-key debounce algorithms would need to be increased too; also doing that on AVR would probably be expensive due to introducing 32-bit operations.

Avoiding multiple timer reads during the debounce processing would be useful (in fact, it should be possible to get the timestamp immediately after the matrix scan and use that timestamp for all key processing during that main loop iteration).

However, this PR seems to become a mix of several independent changes:
- low-level optimizations of timer functions which should not actually change any behavior except maybe making some things faster (or, in some cases, smaller in size but slightly slower);
- test platform enhancements for async tick support;
- debounce algorithm fixes, some of which look like extensive changes.

Those parts should probably be separated (although the new tests for debounce algorithms depend on the test platform changes).

andrebrait commented:
> The problem with making the debounce timers more precise is that the RAM requirements for the per-key debounce algorithms would need to be increased too; also doing that on AVR would probably be expensive due to introducing 32-bit operations.
> 
> Avoiding multiple timer reads during the debounce processing would be useful (in fact, it should be possible to get the timestamp immediately after the matrix scan and use that timestamp for all key processing during that main loop iteration).
> 
> However, this PR seems to become a mix of several independent changes:
> 
> * low-level optimizations of timer functions which should not actually change any behavior except maybe making some things faster (or, in some cases, smaller in size but slightly slower);
> * test platform enhancements for async tick support;
> * debounce algorithm fixes, some of which look like extensive changes.
> 
> Those parts should probably be separated (although the new tests for debounce algorithms depend on the test platform changes).

The "reading the timer twice" issue is fixed now.

As for memory footprint for the per-key stuff, it shouldn't be too much, because the debounce counters can be replaced with small floating point or fixed-point representation, since debounce is, realistically, just some tens of milliseconds. We can get some very good precision still with 8 bits, and excellent with 16-bit. Both of which would be better than skipping a full ms.

Or... We just make it use `>` instead of `>=`, and handle 0 somehow. Then we accept that DEBOUNCE can last up to 1ms longer than it should, but we ensure it never lasts less.

As for the amount of stuff here, yeah, I realize that, but I don't have time to properly split it too much.

I can try to get the main fixes out, but:

1. The increase in size due to the fixes seems to affect AVR boards somewhat, though it's still less than 1% os the size of the microcontrollers I tried out.
2. I can't improve tests without fixing the bugs (since the tests will inherently break) so at best I can get the changes to `sym_defer_g` (if the issue 1 up there isn't a problem) and `sym_defer_pr` into a single PR?
3. Other than some function inlining and some other possible optimizations, there isn't much here that isn't the two things above. Just let me know if they're indeed a problem and I can revert them and leave the rest. However, we're gonna go back to issue number 1.

sigprof commented on platforms/avr/timer.c:
```diff
@@ -111,27 +105,16 @@ inline uint32_t timer_read32(void) {
 /** \brief timer elapsed32
  *
+
  * FIXME: needs doc
  */
 inline uint32_t timer_elapsed32(uint32_t last) {
-    uint32_t t;
-
-    ATOMIC_BLOCK(ATOMIC_RESTORESTATE) {
-        t = timer_count;
-    }
-
-    return TIMER_DIFF_32(t, last);
+    return TIMER_DIFF_32(timer_read32(), last);
 }
```
These functions can actually become generic instead of duplicated for all platforms (their implementation now becomes identical across all platforms).  However, making them actually inlined would probably increase the code size (on AVR a 32-bit comparison is a lot of instructions).

andrebrait replied:
I had made a version like this before, because all implementations were the same, moving the implementation to timer.h (or timer.c, if we don't want to inline them), but later I had one version that kept the AVR code unchanged, but I decided to move forward with this and forgot to deduplicate them.

It didn't make the AVR size larger, though. It stayed the same, IIRC.

sigprof commented on quantum/sync_timer.c:
```diff
@@ -29,31 +29,31 @@ SOFTWARE.
 volatile int32_t sync_timer_ms;
 
-void sync_timer_init(void) {
+inline void sync_timer_init(void) {
```
More pointless `inline`s here.

andrebrait replied:
Ok, I can remove them

sigprof commented on quantum/debounce/tests/sym_defer_g_tests.cpp:
```diff
@@ -236,3 +236,22 @@ TEST_F(DebounceTest, OneKeyDelayedScan4) {
+
+        {10, {}, {{0, 1, UP}}},
+    });
+    /*
+     * Debounce implementations should never read the timer more than once per invocation,
+     * due to some hardware platforms having an asynchronous timer (e.g. ARM).
+     */
```
Actually all hardware platforms have asynchronous timers (`avr` and `arm_atsam` update a millisecond counter from a timer interrupt, `chibios` relies on the ChibiOS system timer, which also gets updated in an asynchronous way).  The only timer implementation that was guaranteeing that the timer value does not change without some explicit action done by the test code was the `test` platform before your changes in this PR.

andrebrait replied:
Yes, I thought that was the case, but I didn't realize it until I dug a little deeper into the code for those platforms. I'll change the comment.

andrebrait commented on platforms/arm_atsam/timer.c:
```diff
@@ -2,34 +2,34 @@
 }
 
-void timer_init(void) {
+inline void timer_init(void) {
```
I have an ARM board which I think uses this, but I can revert all those inlines anyway. 

Also, if declaring as inline in the header is a requirement, none of the AVR ones are getting inlines either.

andrebrait commented on platforms/chibios/timer.c:
```diff
@@ -73,7 +73,7 @@ void timer_clear(void) {
 }
 
-uint16_t timer_read(void) {
+inline uint16_t timer_read(void) {
```
Ok, I can remove them

andrebrait commented:
@sigprof I've redone the PR so it contains *only* the test improvements (+ one slight change to `none`) + fixes for the issues found for `sym_defer_g` and `sym_eager_pr`.

@filterpaper the size for the discussed AVR board went from 16766 bytes to 16768 bytes with these changes now. I basically counter balanced the size increase by restructuring `sym_defer_g` in a way that is more like the other algorithms and it shouldn't make any difference, but somehow smaller on AVR. No idea why, but using a `bool` flag instead of the `uint8_t` counter makes it go from 16766 to 16814 bytes. Well, the increase if largely averted.

filterpaper commented on quantum/debounce/sym_defer_g.c:
```diff
@@ -24,26 +24,39 @@ When no state changes have occured for DEBOUNCE milliseconds, we push the state.
-        if (memcmp(cooked, raw, sizeof(matrix_row_t) * num_rows) != 0) {
-            memcpy(cooked, raw, sizeof(matrix_row_t) * num_rows);
-            cooked_changed = true;
+        debounce_counter = DEBOUNCE;
```
The `debounce_counter` will only change between `DEBOUNCE_ELAPSED` and `DEBOUNCE`. Would a boolean variable (like the original) suffice?

```c
bool debounce(matrix_row_t raw[], matrix_row_t cooked[], uint8_t num_rows, bool changed) {
    static fast_timer_t last_time;
    static bool debouncing = false;
    bool cooked_changed    = false;

    if (changed) {
        debouncing = true;
        last_time  = timer_read_fast();
    } else if (debouncing && timer_elapsed_fast(last_time) >= DEBOUNCE) {
        debouncing         = false;
        size_t matrix_size = num_rows * sizeof(matrix_row_t);
        if (memcmp(cooked, raw, matrix_size) != 0) {
            memcpy(cooked, raw, matrix_size);
            cooked_changed = true;
        }
    }

    return cooked_changed;
}
```

andrebrait replied:
Yes, but then the size of the binary goes back up. I'm not entirely sure of why and what sort of sorcery is behind the compiler's decision on that, but any operations involving booleans make the image a lot larger than using `uint_8` for the same purpose.

And since both data types are equally small and likely use the same pathways inside the controller, I would expect the performance to be the same.

andrebrait replied:
From my other comment:

> No idea why, but using a bool flag instead of the uint8_t counter makes it go from 16766 to 16814 bytes. Well, the increase if largely averted.

filterpaper replied:
In my own userspace (with presence of other and feature), the use of boolean resulted with a smaller firmware. They are likely nuances to optimisation applied with different combination of features and code.

That said, the larger output file difference seem to impact _only_ AVR with _direct pin_ configuration. They are a minority in the repository and I don't think it is worth obfuscating code readability for them.

andrebrait replied:
How much was the difference on the two boards you tested? And what are the command options you're using to compile each?

And I don't think it's obfuscating the readability so much, given it's exactly what the other implementations use as nomenclature and typing. Plus, if it ever becomes necessary to instead decrement the counter in steps, for whatever reason, this code is exactly 3 lines away from doing that. Not that it matters, but I like when every implementation follows the same pattern.

It'll be even more the case after my second PR, which will come after this, because it standardizes and cleans up all implementations.

filterpaper replied:
Just did a quick file size check using `a_dux`:
Current `sym_defer_g`:
```
Checking file size of a_dux_default.hex                                                             [OK]
 * The firmware size is fine - 19996/28672 (69%, 8676 bytes free)
```
Using `debounce_counter_t debounce_counter`:
```
Checking file size of a_dux_default.hex                                                             [OK]
 * The firmware size is fine - 20000/28672 (69%, 8672 bytes free)
```
Using `bool debouncing`:
```
Checking file size of a_dux_default.hex                                                             [OK]
 * The firmware size is fine - 19992/28672 (69%, 8680 bytes free)
```

This is also hard to explain.

andrebrait replied:
My guess is that it has something to do with the DEBOUNCE literal and the comparison with the function return value (alongside some weirdness around using the `bool` flags on the affected boards).

For whatever reason, it seems it's much more prone to optimizing stuff if it sees the same variable storing DEBOUNCE and using it to compare all the other stuff too.

andrebrait replied:
And the way I see it, there's a very, very minor increase in this case, but a "massive" one for the affected boards, so the lesser evil is likely to leave it as `uint8_t`.

filterpaper replied:
> but a "massive" one for the affected boards

I tested with `-kb cradio -km default` and `-kb a_dux -km default` on current `develop` with `avr-gcc 8.5.0`. Both are direct pin PCBs and the boolean variable resulted with more bytes free though.

andrebrait replied:
I'll retest tomorrow, but for cradio, those were the numbers I got.

andrebrait replied:
The difference on the boards that do have an increase with the `uint8_t` counter seem to be minimal (4~8 bytes) whereas for the boards where I saw an increase with the `bool` flag, it was by upwards of 100 bytes, so I think we should keep it this way.

filterpaper replied:
I setup a workflow using `qmkfm/qmk_cli` as a baseline to test different output, using `a_dux` as candidate:

- [develop](https://github.com/filterpaper/qmk_compiler/actions/runs/5958500297/job/16162818277#step:5:96) branch 8674 bytes free
- [sym_defer_g_bool](https://github.com/filterpaper/qmk_compiler/actions/runs/5958500297/job/16162818393#step:5:96) 8678 bytes free (src [diff](https://github.com/filterpaper/qmk_firmware/commit/209492f97f0351ae43000765e446ca9680900940))
- [sym_defer_g_uint8](https://github.com/filterpaper/qmk_compiler/actions/runs/5958500297/job/16162818508#step:5:96) 8670 bytes free (src [diff](https://github.com/filterpaper/qmk_firmware/commit/3c89e3d0a88adb688f2afe2a1b37be7fdf2721dd))

andrebrait replied:
Were all the numbers you sent earlier taken from compiling the code using each file as a custom debounce?

filterpaper replied:
> Were all the numbers you sent earlier taken from compiling the code using each file as a custom debounce?

Yes, they are sourced by the `rules.mk` file and the build action output confirms that the custom `debounce.c` is picked up by the compiler (see the earlier lines in the compile step output). I can see the same results locally but the QMK docker image has the official toolchain and is probably the best neutral resource to validate.

andrebrait replied:
Indeed, now the boolean flag is the smallest of them all, for all boards.

Something happened on `develop` in the meantime though. The firmware is a lot bigger now, in general. Before we were talking about 16.5KB and not it's around 20KB. 

I've reverted it to using a boolean flag now, to recover a few bytes.

Nebuleon commented on quantum/debounce/tests/debounce_test_common.cpp:
```diff
@@ -124,14 +129,20 @@ void DebounceTest::runDebounce(bool changed) {
+    }
+
+    if (current_access_counter() > 1) {
+        FAIL() << "Fatal error: debounce() read the timer multiple times, which is not allowed" << strTime() << "\ntimer: access_count=" << current_access_counter() << "\noutput_matrix: cooked_changed=" << cooked_changed << "\n" << strMatrix(output_matrix_) << "\ncooked_matrix:\n" << strMatrix(cooked_matrix_);
```
Weird output on failure:
> Fatal error: debounce() read the timer multiple times, which is not allowed7812

The suggestion below makes this example output into:
> Fatal error: debounce() read the timer multiple times, which is not allowed, at 7812
```suggestion
        FAIL() << "Fatal error: debounce() read the timer multiple times, which is not allowed, at " << strTime() << "\ntimer: access_count=" << current_access_counter() << "\noutput_matrix: cooked_changed=" << cooked_changed << "\n" << strMatrix(output_matrix_) << "\ncooked_matrix:\n" << strMatrix(cooked_matrix_);
```

andrebrait replied:
Suggestion applied. Thanks!

Nebuleon reviewed (approved):
I've tested these changes to `none`, `sym_defer_g` and `sym_eager_pr` on my keyboard while playing a Touhou MIDI file (notoriously full of notes) to induce extra interrupts taking 1-3 ms every 7.8 ms at `DEBOUNCE` values of 1 and 5, as well as disabling my per-key RGB to increase the matrix scan rate, and the changes work at least as well as the original code.

I've also looked at the new `AsyncTickOneKeyShort1` test and the changes it required to the test harness, and they look good.

In order to make the tests you introduce in this PR pass for `sym_defer_g`, you had to make some changes to it, and those are also good. The changes to `sym_eager_pr` to make sure `cooked_changed` is updated based on the correct cooked row, i.e. before the update, are good and needed as well.

The only thing that remains is the matter of `sym_defer_g` being a bit larger on AVR. But, since it is now  only calling the timer function once, which is more correct in the face of interrupts updating the timer, it's not a concern for me if this happens to be a few bytes larger.

andrebrait commented:
> The only thing that remains is the matter of `sym_defer_g` being a bit larger on AVR. But, since it is now only calling the timer function once, which is more correct in the face of interrupts updating the timer, it's not a concern for me if this happens to be a few bytes larger.

@Nebuleon not anymore. It's now smaller than in develop. @filterpaper these are the latest results.

* develop:
```
a_dux:
 * The firmware size is fine - 19996/28672 (69%, 8676 bytes free)

cradio:
 * The firmware size is fine - 19974/28672 (69%, 8698 bytes free)
```

* this PR:
```
a_dux: 
 * The firmware size is fine - 19988/28672 (69%, 8684 bytes free)

cradio:
 * The firmware size is fine - 19966/28672 (69%, 8706 bytes free)
```

drashna commented on quantum/debounce/sym_defer_g.c:
```diff
@@ -24,6 +24,12 @@ When no state changes have occured for DEBOUNCE milliseconds, we push the state.
+// Maximum debounce: 255ms
+#if DEBOUNCE > UINT8_MAX
+#    undef DEBOUNCE
+#    define DEBOUNCE UINT8_MAX
```
Not sure it's the best option, but maybe a pragma message should be here to notify the user, if not an error outright.

andrebrait replied:
Sure. I'm just following what the other implementations do here.

I can make that a general improvement in the PR that will follow.

drashna replied:
Ah, yeah, the other files are doing the same.  In that case, it should be okay here, as I'd rather it be consistent.

Nebuleon reviewed (approved):
All OK for me. The change to warn on overflow of `#define DEBOUNCE` instead of clamping in all implementations can go in another PR, which would conflict with this one so it must be done after this one is merged.

andrebrait commented:
Any chance we can get this merged? @drashna @sigprof @filterpaper 

I think it was reduced enough and tested enough to be proven benign. I have a couple other improvements/cleanups on the way but those depend on this being on `develop` at least.
